File: basic/1213/1213.py
import sys
from collections import deque

cnt = [0 for _ in range(26)]
flag = 0
mid = ''
ans = deque([])

s = sys.stdin.readline().strip()
for c in s:
    cnt[ord(c) - ord('A')] += 1

for i in range(len(cnt)-1, -1, -1):
    if cnt[i] % 2 == 1:
        flag += 1
        mid = chr(i+ord('A'))
        if flag == 2:
            break
            
    for j in range(cnt[i]//2):
        ans.append(chr(i+ord('A')))
        ans.appendleft(chr(i+ord('A')))

if flag == 2:
    sys.stdout.write("I'm Sorry Hansoo"+"\n")
else:
    ans_list = list(ans)
    if len(mid):
        ans_list.insert(len(ans_list)//2, mid)
    sys.stdout.write(''.join(ans_list)+"\n")

File: baekjoon/basic1/1213/1213.cpp
#include <bits/stdc++.h>
using namespace std;

string s, ret;
char mid;
int flag;
int cnt[200];

int main(){
    ios_base::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);

    cin >> s;
    for (int c : s) cnt[c]++;
    for(int i='Z'; i>='A'; i--){
        if(cnt[i]){
            if (cnt[i] % 2 == 1){
                flag++;
                mid = char(i);
                cnt[i] --;
            }
            if (flag == 2) break;
            for(int j = 0; j < cnt[i]; j+=2){
                ret = char(i) + ret;
                ret += char(i);
            }
        }
    }
    if (flag == 2) cout << "I'm Sorry Hansoo\n";
    else{
        if (mid) ret.insert(ret.begin() + ret.size()/2, mid);
        cout << ret << "\n";
    }
    return 0;
}